snapshots/windows/windows.go

```diff
@@ -193,17 +201,72 @@ func (s *snapshotter) Commit(ctx context.Context, name, key string, opts ...snap
 	// grab the existing id
 	id, _, _, err := storage.GetInfo(ctx, key)
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "failed to get storage info for %s", key)
 	}
 
-	usage, err := fs.DiskUsage(ctx, s.getSnapshotDir(id))
+	snapshot, err := storage.GetSnapshot(ctx, key)
 	if err != nil {
 		return err
 	}
 
+	path := s.getSnapshotDir(id)
+
+	// In order to turn a Scratch Layer (sandbox.vhdx with all the data) into a
+	// Read-only Layer (Files/, Hives/, etc) we need to create a new snapshot and
+	// restream the data out of the old snapshot.
+
+	// Some Snapshots will already have Files/ etc, if they were created by streaming in a layer
+	// archive. Layers created during image pull will be in this state, for example. In that case,
+	// we should delete sandbox.vhdx if still present, as it will not reflect the layer content.
+	// A layer with an empty parent list will never have had a sandbox.vhdx.
+
+	// TODO: Is this the right test? What we want is whatever causes a layer to fail when included
+	// in the parentpaths for another layer with an 0x2 file-not-found failure.
+	if _, stErr := os.Stat(filepath.Join(path, "Files")); stErr != nil && !os.IsNotExist(stErr) {
```

Comment:
Per https://github.com/microsoft/hcsshim/issues/853, it looks like we can assume that a layer with no parents has a Files/ directory, and if it doesn't have /layout then it needs `hcsshim.ProcessBaseLayer` called on it, and may need `hcsshim.ProcessUtilityVMImage` called too. These two functions are called internally by hcsshim when extracting a base-layer tarball, and I might add an API to hcsshim instead of here to call them if necessary, since the logic of "is necessary" already lives in hcsshim, in `baseLayerWriter`'s implementation.

This means this test changes to checking the count of `snapshot.ParentIDs`.

It's a small change here, but flows on to changes in in #4419, as bind-mounts for active parentless layers must bind-mount the _Files/_ directory under the layer source. As such, I'm marking both this and #4419 as draft until I get around to doing those changes.

Full credit to #2366 which actually had implemented the necessary support for "create a base layer", but as part of the integration testsuite, which I only looked closely at today.

Reply:
After #4643 we can probably invert this to check for the positive presence of sandbox.vhdx, in which case we need to restream, as we should never be in the position of having both sandbox.vhdx and Files/ present at `Commit`-time.

Reply:
https://github.com/microsoft/hcsshim/pull/901 is the proposed feature for hcsshim on which I plan to rebase and simplify this and #4419. It's basically functional, but has not yet been reviewed. It's good enough that next time I play with this, I can start changing this PR to use it, and ideally end up with a demonstrably-working stack that passes the snapshotter test suite on Windows without needing to significantly _change_ the suite, as #2366 needed to.